sdk/src/azure/iot/az_iot_adu_client.c

```diff
@@ -0,0 +1,872 @@
+        _az_RETURN_IF_FAILED(az_json_writer_append_string(
+            ref_json_writer, last_install_result->step_results[i].result_details));
+      }
+
+      _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+    }
+
+    _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+  }
+
+  /* Fill the agent state.   */
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+      ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_STATE)));
+  _az_RETURN_IF_FAILED(az_json_writer_append_int32(ref_json_writer, agent_state));
+
+  /* Fill the workflow.  */
+  if (workflow != NULL && (az_span_ptr(workflow->id) != NULL && az_span_size(workflow->id) > 0))
+  {
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_WORKFLOW)));
+    _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(ref_json_writer));
+
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_ACTION)));
+    _az_RETURN_IF_FAILED(az_json_writer_append_int32(ref_json_writer, workflow->action));
+
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+        ref_json_writer, AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_ID)));
+    _az_RETURN_IF_FAILED(az_json_writer_append_string(ref_json_writer, workflow->id));
+
+    /* Append retry timestamp in workflow if existed.  */
+    if (!az_span_is_content_equal(workflow->retry_timestamp, AZ_SPAN_EMPTY))
+    {
+      _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+          ref_json_writer,
+          AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_RETRY_TIMESTAMP)));
+      _az_RETURN_IF_FAILED(
+          az_json_writer_append_string(ref_json_writer, workflow->retry_timestamp));
+    }
+    _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+  }
+
+  /* Fill installed update id. */
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(
+      ref_json_writer,
+      AZ_SPAN_FROM_STR(AZ_IOT_ADU_CLIENT_AGENT_PROPERTY_NAME_INSTALLED_UPDATE_ID)));
+  _az_RETURN_IF_FAILED(az_json_writer_append_string(ref_json_writer, device_properties->update_id));
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_end_component(NULL, ref_json_writer));
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(ref_json_writer));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_adu_client_parse_service_properties(
+    az_iot_adu_client* client,
+    az_json_reader* ref_json_reader,
+    az_iot_adu_client_update_request* update_request)
+{
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(ref_json_reader);
+  _az_PRECONDITION_NOT_NULL(update_request);
+
+  (void)client;
```

Comment:
Almost all functions seem to be related to a JSON reader.
Looks like it may be a good idea to move the reader as a private member of the `client` instead of a parameter to the function.

Reply:
We can't (or shouldn't). The user has to initialize the JSON reader, which allows Azure RTOS to initialize it with discontinuous memory.

Reply:
Just to note, this follows the precedent already set with PnP APIs